### ISSUE TITLE
How to prune own models

### ISSUE THREAD

**Issue by mskmei** [ziplab/SPViT#2]
Hey, thanks for the wonderful work and amazing code!
I wonder whether we can modify this code to let it work on own models, like basic ViT from vit_pytorch. Can you give some simple instructions? Thank you!

**Comment by Charleshhy** [comment]
Hi mskmei,

Thanks for your interest in our work and I am terribly sorry for the late reply (I forgot to watch this project...).

To prune your own model, you need to customize the `UnifiedAttention` and `UnifiedMlp` layers and customize the FLOPs calculation function similar to [here](https://github.com/ziplab/SPViT/blob/master/SPViT_DeiT/models_pruning.py#L793). However, if you wish to prune vit_pytorch, I believe you can directly override `Attention` and `FeedForward` layers with our implementations in SPViT_DeiT.

Kind regards,
Haoyu

**Comment by buibui22** [comment]
请问你们的categories.json和train.json里面的格式是什么样子

**Comment by Charleshhy** [comment]
> 请问你们的categories.json和train.json里面的格式是什么样子

I assume you refer to [this line](https://github.com/ziplab/SPViT/blob/29dd0acd4ce2471225b6445dcdd4cb2942edf022/SPViT_DeiT/datasets.py#L27). Actually this part of code for training iNaturalist is borrowed from [DeiT repository](https://github.com/facebookresearch/deit/blob/main/datasets.py) and I have not tested it in our project. You may want to find the format of `categories.json` and `train.json` there.

**Comment by buibui22** [comment]
&nbsp;categories.json&nbsp;and&nbsp;train.json格式是什么，可以发给我吗，十分感谢




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "ziplab/SPViT"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年11月2日(星期四) 上午10:40
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [ziplab/SPViT] How to prune own models (Issue #2)





  
请问你们的categories.json和train.json里面的格式是什么样子
  
I assume you refer to this line. Actually this part of code for training iNaturalist is borrowed from DeiT repository and I have not tested it in our project. You may want to find the format of categories.json and train.json there.

**Comment by buibui22** [comment]
用自己的数据集跑的话应该修改哪些部分
------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "ziplab/SPViT"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年11月2日(星期四) 上午10:40
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [ziplab/SPViT] How to prune own models (Issue #2)





  
请问你们的categories.json和train.json里面的格式是什么样子
  
I assume you refer to this line. Actually this part of code for training iNaturalist is borrowed from DeiT repository and I have not tested it in our project. You may want to find the format of categories.json and train.json there.

**Comment by Charleshhy** [comment]
> 用自己的数据集跑的话应该修改哪些部分 ------------------&nbsp;原始邮件&nbsp;------------------ 发件人: "ziplab/SPViT" ***@***.***&gt;; 发送时间:&nbsp;2023年11月2日(星期四) 上午10:40 ***@***.***&gt;; ***@***.******@***.***&gt;; 主题:&nbsp;Re: [ziplab/SPViT] How to prune own models (Issue #2) 请问你们的categories.json和train.json里面的格式是什么样子 I assume you refer to this line. Actually this part of code for training iNaturalist is borrowed from DeiT repository and I have not tested it in our project. You may want to find the format of categories.json and train.json there. — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you commented.Message ID: ***@***.***&gt;

Our tested datasets are CIFAR100 and ImageNet. Basically, you can align the format of your own dataset to CIFAR100 or Imagenet and use their loading code.

**Comment by GreedyZJH** [comment]
> Hi mskmei,
> 
> Thanks for your interest in our work and I am terribly sorry for the late reply (I forgot to watch this project...).
> 
> To prune your own model, you need to customize the `UnifiedAttention` and `UnifiedMlp` layers and customize the FLOPs calculation function similar to [here](https://github.com/ziplab/SPViT/blob/master/SPViT_DeiT/models_pruning.py#L793). However, if you wish to prune vit_pytorch, I believe you can directly override `Attention` and `FeedForward` layers with our implementations in SPViT_DeiT.
> 
> Kind regards, Haoyu

Hello, I would like to know whether it is necessary to distill knowledge using the teacher-student structure if only the FNN layer is pruned?

**Comment by Charleshhy** [comment]
> > Hi mskmei,
> > Thanks for your interest in our work and I am terribly sorry for the late reply (I forgot to watch this project...).
> > To prune your own model, you need to customize the `UnifiedAttention` and `UnifiedMlp` layers and customize the FLOPs calculation function similar to [here](https://github.com/ziplab/SPViT/blob/master/SPViT_DeiT/models_pruning.py#L793). However, if you wish to prune vit_pytorch, I believe you can directly override `Attention` and `FeedForward` layers with our implementations in SPViT_DeiT.
> > Kind regards, Haoyu
> 
> Hello, I would like to know whether it is necessary to distill knowledge using the teacher-student structure if only the FNN layer is pruned?

Hi GreedyZJH,

I have not conducted this experiment but I do believe knowledge distillation can lead to better performance!

Best,
Haoyu

**Comment by GreedyZJH** [comment]
> > > Hi mskmei,
> > > Thanks for your interest in our work and I am terribly sorry for the late reply (I forgot to watch this project...).
> > > To prune your own model, you need to customize the `UnifiedAttention` and `UnifiedMlp` layers and customize the FLOPs calculation function similar to [here](https://github.com/ziplab/SPViT/blob/master/SPViT_DeiT/models_pruning.py#L793). However, if you wish to prune vit_pytorch, I believe you can directly override `Attention` and `FeedForward` layers with our implementations in SPViT_DeiT.
> > > Kind regards, Haoyu
> > 
> > 
> > Hello, I would like to know whether it is necessary to distill knowledge using the teacher-student structure if only the FNN layer is pruned?
> 
> Hi GreedyZJH,
> 
> I have not conducted this experiment but I do believe knowledge distillation can lead to better performance!
> 
> Best, Haoyu
Hi Haoyu,
Thinks for you answer. I am a researcher in a bioinformatics related field and am studying your article. I may not have phrased my question clearly. I'm a bit confused about the learnable parameters of the FNN layer, which uses the Bernoulli distribution. But when I apply it to own Transformer model, the loss_bop loss function always fluctuates and maintains in an interval, showing disorder. It feels like there is something wrong with my model. I just applied your UnifiedFFN code, is there something I didn't take into account?
Best, GreedyZJH

**Comment by Charleshhy** [comment]
> > > > Hi mskmei,
> > > > Thanks for your interest in our work and I am terribly sorry for the late reply (I forgot to watch this project...).
> > > > To prune your own model, you need to customize the `UnifiedAttention` and `UnifiedMlp` layers and customize the FLOPs calculation function similar to [here](https://github.com/ziplab/SPViT/blob/master/SPViT_DeiT/models_pruning.py#L793). However, if you wish to prune vit_pytorch, I believe you can directly override `Attention` and `FeedForward` layers with our implementations in SPViT_DeiT.
> > > > Kind regards, Haoyu
> > > 
> > > 
> > > Hello, I would like to know whether it is necessary to distill knowledge using the teacher-student structure if only the FNN layer is pruned?
> > 
> > 
> > Hi GreedyZJH,
> > I have not conducted this experiment but I do believe knowledge distillation can lead to better performance!
> > Best, Haoyu
> > Hi Haoyu,
> > Thinks for you answer. I am a researcher in a bioinformatics related field and am studying your article. I may not have phrased my question clearly. I'm a bit confused about the learnable parameters of the FNN layer, which uses the Bernoulli distribution. But when I apply it to own Transformer model, the loss_bop loss function always fluctuates and maintains in an interval, showing disorder. It feels like there is something wrong with my model. I just applied your UnifiedFFN code, is there something I didn't take into account?
> > Best, GreedyZJH

Sorry for the late reply. It's normal to have fluctuated bop, as we jointly optimize the cross-entropy loss and loss_bop (which aims to optimize the gate parameters and achieve your target FLOPs) and it's challenging to get to a sweet point for both losses. Since sometimes loss_bop cannot converge well to a stable minimum, you can try early stopping or put higher weights to loss_bop and make sure it would converge.

**Comment by bo102** [comment]
> 己的模型，您需要自定义 和 层，并自定义 FLO

 you can send me a swin-transformer-tiny model that you pruned with your pruning method, thank you very much

**Comment by bo102** [comment]
> 嗨，mskmei，
> 
> 感谢您对我们工作的关注，对于迟到的回复，我感到非常抱歉（我忘了看这个项目......
> 
> 要修剪自己的模型，您需要自定义 和 层，并自定义 FLOPs 计算函数，类似于[此处](https://github.com/ziplab/SPViT/blob/master/SPViT_DeiT/models_pruning.py#L793)。但是，如果您想修剪vit_pytorch，我相信您可以直接覆盖和分层我们的 SPViT_DeiT 实现。`UnifiedAttention``UnifiedMlp``Attention``FeedForward`
> 
> 亲切的问候，浩宇
Hello, do you have the FPS after pruning the model running on imagenet-1k?

**Comment by bo102** [comment]
Can you send me a swin-transformer-tiny model file with the model structure and weights after pruning? (Only pruning, not knowledge distillation) Thank you very much

**Comment by Charleshhy** [comment]
Close this issue for better clarification. Please create new issues if your question is not directly related to the original topic :)
